File: 2021/Day05/Day5.scala
import scala.io.Source

@main def day5 =
    type Point = (Int, Int)
    type Start = Point
    type End = Point
    type Solution = ((Start, End)) => List[Point]

    def range(a: Int, b: Int) =
        if a > b then a to b by -1
        else a to b

    def straight(start: Start, end: End) =
        val ((x1, y1), (x2, y2)) = (start, end)
        (
          for x <- range(x1, x2)
          yield for y <- range(y1, y2)
          yield (x, y)
        ).flatMap(_.toList).toList

    def diagonal(start: Start, end: End) =
        val ((x1, y1), (x2, y2)) = (start, end)
        (range(x1, x2) zip range(y1, y2)).toList

    def firstPart(start: Start, end: End) =
        val ((x1, y1), (x2, y2)) = (start, end)
        if x1 != x2 && y1 != y2 then Nil
        else straight(start, end)

    def secondPart(start: Start, end: End) =
        val ((x1, y1), (x2, y2)) = (start, end)
        if x1 != x2 && y1 != y2 then diagonal(start, end)
        else straight(start, end)

    def input = Source
        .fromFile("input.txt")
        .getLines
        .map(_.split(" -> "))
        .map(line =>
            val Array(x1, y1) = line(0).split(",").map(_.toInt)
            val Array(x2, y2) = line(1).split(",").map(_.toInt)
            ((x1, y1), (x2, y2))
        )
        .toList

    def solve(input: List[(Start, End)], solution: Solution) =
        input
            .map(solution)
            .flatMap(identity)
            .groupBy(identity)
            .view
            .mapValues(_.size)
            .toMap
            .count(_._2 > 1)

    println(s"First part: ${solve(input, firstPart)}")
    println(s"Second part: ${solve(input, secondPart)}")

File: 2021/archived/Day05/Day5.fs
module Day5

open System
open System.IO

let t2 a =
    match a with
    | [| x; y |] -> (x, y)
    | _ -> failwithf "Bang %A" a

let step a b = if a < b then 1 else -1

let straight (x1, y1) (x2, y2) =
    [| for x in x1 .. step x1 x2 .. x2 do
           for y in y1 .. step y1 y2 .. y2 -> (x, y) |]

let diagonal (x1, y1) (x2, y2) =
    Array.zip [| x1 .. step x1 x2 .. x2 |] [| y1 .. step y1 y2 .. y2 |]

let fisrstPart ((x1, y1), (x2, y2)) =
    if x1 <> x2 && y1 <> y2 then
        [||]
    else
        straight (x1, y1) (x2, y2)

let secondPart ((x1, y1), (x2, y2)) =
    if x1 <> x2 && y1 <> y2 then
        diagonal (x1, y1) (x2, y2)
    else
        straight (x1, y1) (x2, y2)


let solve input solution =
    input
    |> Array.map solution
    |> Array.concat
    |> Array.countBy id
    |> Array.filter (fun (_, count) -> count > 1)
    |> Array.length

let input =
    File.ReadAllLines "input.txt"
    |> Array.map (fun s -> s.Split([| " -> " |], StringSplitOptions.None))
    |> Array.map (fun pos ->
        let parse = (Array.map int) >> t2

        (pos.[0].Split [| ',' |] |> parse, pos.[1].Split [| ',' |] |> parse))

solve input fisrstPart |> printfn "First part: %d"
solve input secondPart |> printfn "First part: %d"
